### ISSUE TITLE
AvroConsumer is slow

### ISSUE THREAD

**Issue by ajagarapu** [confluentinc/confluent-kafka-python#346]
Description
===========
I am using the AvroConsumer to read from a topic with 64 partitions and i see that the consumer has a throughput of processing close to 500 messages per second. Here is the profiler for the code and basically it consumed around 100,000 messages and it took a long time.

Wed Mar 28 20:58:28 2018    profiler_1

         1149855327 function calls (1089514249 primitive calls) in 929.361 seconds

   Ordered by: internal time
   List reduced from 2193 to 30 due to restriction <30>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
59342690/230906  280.142    0.000  841.719    0.004 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:483(read_data)
 82623553  127.433    0.000  166.420    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:414(match_schemas)
549874829  123.711    0.000  123.711    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/schema.py:238(type)
   230906   63.936    0.000  836.243    0.004 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:700(read_record)
 19165148   57.412    0.000  471.704    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:675(read_union)
 32366191   45.430    0.000   70.678    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:162(read)
 24601984   41.353    0.000  105.525    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:194(read_long)
 57495441   14.988    0.000   14.988    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/schema.py:893(schemas)
   244985   14.555    0.000   14.555    0.000 {function AvroConsumer.poll at 0x7effc132a9d8}
    86762   13.099    0.000   13.099    0.000 {method 'recv' of '_socket.socket' objects}
 41563159   12.218    0.000   12.218    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/schema.py:588(name)
52817857/52817618   12.032    0.000   12.033    0.000 {built-in method builtins.len}
 41562980   11.690    0.000   11.690    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/schema.py:583(type)
 32597097    9.713    0.000    9.713    0.000 {method 'read' of '_io.BytesIO' objects}
 32366191    8.345    0.000    8.345    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:157(reader)
  5039085    6.819    0.000   41.840    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:236(read_bytes)
 21095543    6.598    0.000    6.598    0.000 {method 'get' of 'dict' objects}
  5039085    6.028    0.000   50.861    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:244(read_utf8)
 27327877    5.625    0.000    5.625    0.000 {built-in method builtins.ord}
1755926/877963    4.123    0.000   15.457    0.000 /usr/lib64/python3.6/configparser.py:765(get)
 15049431    3.553    0.000    3.553    0.000 /home/ec2-user/Apps/lib/python3.6/site-packages/avro/io.py:175(read_null)
        1    3.040    3.040  929.385  929.385 middle-man.py:1(<module>)
  5053981    3.006    0.000    3.006    0.000 {method 'decode' of 'bytes' objects}
  1755926    2.539    0.000    3.888    0.000 /usr/lib64/python3.6/configparser.py:1131(_unify_values)
  1755926    2.210    0.000    2.210    0.000 /home/ec2-user/Apps/lib64/python3.6/collections/__init__.py:877(__getitem__)
    43381    2.179    0.000    2.759    0.000 {built-in method pymongo._cmessage._query_message}
    21690    2.019    0.000   22.837    0.001 middle-man.py:21(pruneData)
   877963    1.939    0.000    7.852    0.000 /usr/lib64/python3.6/configparser.py:405(_interpolate_some)
   230906    1.844    0.000  846.375    0.004 /home/ec2-user/Apps/lib64/python3.6/site-packages/confluent_kafka/avro/serializer/message_serializer.py:198(decode_message)
   877963    1.720    0.000   18.996    0.000 /usr/lib64/python3.6/configparser.py:1231(__getitem__)

How to reproduce
================

code:
from confluent_kafka import KafkaError,KafkaException
from confluent_kafka.avro import AvroConsumer
<config code>
consumer = AvroConsumer({
            'bootstrap.servers':config['kafka']['brokers'],
            'group.id' : config['kafka']['group.id'],
            'schema.registry.url' : config['kafka']['schema.registry.url'],
            'socket.keepalive.enable' : True})

consumer.subscribe(['my_topic'])
running = True
try:
  while running:
   msg = consumer.poll(timeout=0.001)
   if msg is None:
    continue
   if msg.error():
    if(msg.error().code() == KafkaError._PARTITION_EOF):
     logger.error('%% %s [%d] End of the partition %d\n' % (msg.topic(), msg.partition(), msg.offset()))
   else:
    print(msg.value())
except KeyboardInterrupt:
 sys.stderr.write('%% Aborted by user\n')
finally:
 consumer.close()

Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`): confluent_kafka.version(): ('0.11.0', 720896)
 and confluent_kafka.libversion(): ('0.11.3', 721919)
 - [x] Apache Kafka broker version: 0.11.0
 - [x] Client configuration: `{
            'bootstrap.servers':config['kafka']['brokers'],
            'group.id' : config['kafka']['group.id'],
            'schema.registry.url' : config['kafka']['schema.registry.url'],
            'socket.keepalive.enable' : True}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by c-nichols** [comment]
I noticed today that fastavro raises an exception while deserializing.  This might be the cause of the slowness, since it falls back to the pure-python version.

The root cause seems to have someting to do with the avro schema being passed in to fastavro.

Versions:
- confluent-kafka master branch HEAD
- fastavro==0.17.9
- confluent platform v4.0.0

**Comment by edenhill** [comment]
@c-nichols I suggest opening an issue with the fastavro project, providing a reproducible test case, to get this fixed.

**Comment by glgnohk** [comment]
Using 

- Python 3.5.2
- confluent-kafka==0.11.4
- fastavro==0.17.9

I got the same issue. 

Trying a few versions of fastavro back to 0.14.11, I still see the issue. 

I also tried avro-python3 on 1.8.2 (latest), 1.80 and 1.77 and still see this issue. I also didn't see changes to to_json() code in avro.schema.RecordSchema, which is the type right before passed to fastavro (https://github.com/confluentinc/confluent-kafka-python/blob/40a5b3e70044791cf5384ba3a7a0b1ab24450381/confluent_kafka/avro/serializer/message_serializer.py#L171) So not sure when ```avro``` switched to ImmutableDict.

Anyway, when I did confluent-kafka==0.11.0, then we don't get this error. I don't see this on  confluent-kafka==0.9.4 either. So this appears like it could be a regression in confluent-kafka

Final versions that works fine:

- avro-python3==1.8.2 (latest)
- fastavro==0.17.10 (latest)
- confluent-kafka==0.11.0 (previous version)

**Comment by qiao-meng-zefr** [comment]
I'm having the exact same issue. Here are my environment:
```
Python 3.6.1
avro-python3==1.8.2
confluent-kafka==0.11.4
fastavro==0.17.10
```

Don't really want to roll back to confluent-kafka 0.11.0 because some nice feature in 0.11.4..

Haven't find a workaround yet.

**Comment by scottbelden** [comment]
It seems like the `ImmutableDict` from `avro-python3` is getting passed to `fastavro` as the schema and it breaks at the cython level because an `ImmutableDict` is not a `dict`. If the schema can be turned into a normal dict before being passed to `fastavro` then that should resolve this, but if not then fastavro could remove some of the expectations of getting a simple dict and accept a dict-like object. I don't think there would be much/any performance hit as there are only a few accesses of the dictionary: https://github.com/tebeka/fastavro/blob/master/fastavro/_schema.pyx#L24-L26

**Comment by scottbelden** [comment]
I don't have kafka set up, but if someone is able to run with [this branch of fastavro](https://github.com/tebeka/fastavro/compare/isinstance_not_type) then could you let me know if that resolves the issue. If so I can get that merged in.

**Comment by c-nichols** [comment]
I ran a quick (unscientific but good enough for comparison) benchmark against 100k records

With fastavro==0.19.0:
```
Processed 100000 records in 16.697677612304688 seconds (5988.856793253032/sec)
```

With your branch:
```
Processed 100000 records in 2.8512444496154785 seconds (35072.404968113515/sec)
```

I don't have context on what the numbers *should* be, but that looks like a pretty massive improvement  :) :+1:

**Comment by scottbelden** [comment]
Thanks! I'll get the fix merged and released shortly.

**Comment by scottbelden** [comment]
@c-nichols I applied the fix in 0.19.1. Thanks for the help with diagnosing!

**Comment by edenhill** [comment]
Whoa, impressive performance improvement! :1st_place_medal:
